phare/pp/diagnostics.py

```diff
@@ -0,0 +1,234 @@
+        return Diagnostics.get(self.diags, "ions_pop_ions_${POP}_density")
+
+    def getDensity(self, pop=""):
+        if len(pop):
+            return Diagnostics.get(self.diags, "ions_pop_ions_" + pop + "_density")[0]
+        return Diagnostics.get(self.diags, "ions_density")[0]
+
+    @staticmethod
+    def _extract_recursive(diag, h5item, pre_key=""):
+        assert isinstance(diag, Diagnostic)
+
+        kinder = list(h5item.keys())
+        n_keys = len(kinder)
+
+        if n_keys == 1 and isinstance(h5item[kinder[0]], h5py.Dataset):
+            """Single dataset per patch hdf5 file, e.g. ion density"""
+            return diag.patch_data_type(diag, kinder[0], h5item[kinder[0]])
+
+        if n_keys > 1 and isinstance(h5item[kinder[0]], h5py.Dataset):
+            """Multiple dataset per patch hdf5 file, e.g. ion flux"""
+            return diag.patch_data_type(diag, pre_key, h5item)
+
+        if n_keys:
+            return Diagnostics._extract_recursive(diag, h5item[kinder[0]], kinder[0])
+
+    @staticmethod
+    def _extract(diag, h5item):
+        assert isinstance(diag, Diagnostic)
+        return Diagnostics._extract_recursive(diag, h5item)
+
+    @staticmethod
+    def _get_h5_files(sim: ph.Simulation):
+        assert isinstance(sim, ph.Simulation)
+
+        h5files = []
+        for root, dirs, files in os.walk(sim.diag_options["options"]["dir"]):
+            for file in files:
+                if Diagnostic.is_input_file_valid(file):
+                    h5files.append(os.path.join(root, file))
+                else:
+                    print("Warning: Input file not recognized, ignored: ", file)
+        return h5files
+
+    @staticmethod
+    def _validate_simulator(sim: ph.Simulation):
+        """Fails if the input Simulation object has no diagnostic output"""
+
+        assert isinstance(sim, ph.Simulation)
+
+        if any(
+            [
+                sim.diag_options is None,
+                sim.diag_options["options"] is None
+                or sim.diag_options["options"]["dir"] is None,
+            ]
+        ):
+            raise ValueError(
+                "Cannot load diagnostics from specific simulation: no diagnostics made"
+            )
+
+    @staticmethod
+    def _merge_particle_pops(diags):
+        """Takes per HDF5 file Diagnostic objects and coallates the particle population
+            types into one DType object."""
+
+        ndiags, particle_pops = [], {}
+        for d in diags:
+            if not isinstance(d.type, _Particle):
+                ndiags.append(d)
+            else:
+                if d.type.pop not in particle_pops:
+                    particle_pops[d.type.pop] = {}
+                particle_pops[d.type.pop][d.type.name] = d
+        for p, t in particle_pops.items():
+            ndiags.append(Particles(p, t["domain"], t["patchGhost"], t["levelGhost"]))
+        return ndiags
+
+    @staticmethod
+    def extract(sim: ph.Simulation):
```

Comment:
sorry I don't understand what this function does.
you make a list  of diagnotic from you h5 file list.
then you make a dict of ... what? you are looping over _EM, _Fluid and _Particle but you only call a method called `_merge_paticle_pops` which thus implies (just by the name) not to deal with EM.
So I guess that's merge particle pops that is badly named...

also I find it confusing that Extract does not call the _extrat* routines. what are they for?

Reply:
at this point we don't have the {PatchDataType : diagnostic_list} dict, so we create it

in order to later deal with the three separate h5 files for particles easily, we merge the domain. patchGhost, and levelGhost diagnostics into one object. but we also remove the no longer needed diagnostics of domain. patchGhost, and levelGhost as they are encapsulated in another type.

Reply:
ok. what I just find confusing is that you have a dict of EM, Particle and Fluid to which you apply the same treatment, from the extract function point of view.
But that treatment actually only applies to _Particles since in that function there is 

```
            if not isinstance(d.type, _Particle):
                ndiags.append(d)
```
that does not read well.

Reply:
it's only used for particle overlaps, if someone needs it they can build it themselves
will reassess